### ISSUE TITLE
Amazon Prime Gaming sync failure duo to TypeError "NoneType is not iterable"

### ISSUE THREAD

**Issue by Gamerpunk25** [lutris/lutris#5596]
### Bug description

After i added as source for games amazon prime gaming and logged into my amazon account, i can see on the side, that the sync try is in process. After a few seconds, the log output opens and i get to see the error message below in the section "Log output"
I am using the current flatpack version of lutris and let my pc search for any updates via the program "Discover"

### How to Reproduce

1. Start lutris via the shortcut on my desktop
2. Go in the section "source" to amazon prime gaming and hit the sync icon
3. Wait for the response because i am already logged in
4. Log output opens with the error message

### Expected behavior

Overview of my Games on my Prime Gaming account

### Log output

```shell
'NoneType' object is not iterable

Traceback (most recent call last):

  File "/app/lib/python3.11/site-packages/lutris/util/jobs.py", line 30, in target
    result = self.function(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/app/lib/python3.11/site-packages/lutris/services/base.py", line 142, in do_reload
    self.load()

  File "/app/lib/python3.11/site-packages/lutris/services/amazon.py", line 160, in load
    games = [AmazonGame.new_from_amazon_game(game) for game in self.get_library()]
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

TypeError: 'NoneType' object is not iterable
```

### System Information

```shell
Vulkan support: YES
Esync support: YES
Fsync support: YES
Wine installed: YES
Gamescope: NO
Mangohud: NO
Gamemode: YES
Steam: YES
In Flatpak: YES
[System]
OS: GNOME 45 Flatpak runtime
Arch: x86_64
Kernel: 6.8.0-101041-tuxedo
Desktop: KDE
Display Server: wayland
[CPU]
Vendor: AuthenticAMD
Model: AMD Ryzen 7 7840HS w/ Radeon 780M Graphics
Physical cores: 8
Logical cores: 16
[Memory]
RAM: 14.8 GB
Swap: 15.9 GB
[Graphics]
Vendor: AMD
OpenGL Renderer: AMD Radeon 780M (radeonsi, gfx1103_r1, LLVM 17.0.6, DRM 3.57, 6.8.0-101041-tuxedo)
OpenGL Version: 4.6 (Compatibility Profile) Mesa 24.1.3 (git-0c49f54c76)
OpenGL Core: 4.6 (Core Profile) Mesa 24.1.3 (git-0c49f54c76)
OpenGL ES: OpenGL ES 3.2 Mesa 24.1.3 (git-0c49f54c76)
Vulkan Version: 1.3.261
Vulkan Drivers: AMD Radeon 780M (RADV GFX1103_R1) (1.3.278), NVIDIA GeForce RTX 4060 Laptop GPU (1.3.277)
```

### Media (optional)

![grafik](https://github.com/user-attachments/assets/bf961ae5-1d99-4053-a7f9-a579c46a85bd)

**Comment by Gamerpunk25** [comment]
Alright danieljohnson2, i'll wait for the notification for further instructions

**Comment by danieljohnson2** [comment]
Okay, @Gamerpunk25, let's give this a try.

I want you to download [this branch](https://github.com/lutris/lutris/tree/dj/amazon_error_handling) and then run the ``bin/lutris`` script within. No need to install anything.

Try to reload your Amazon source; I expect you will get an error, and I would like to see that error.

I expect the error to be displayed to you, but also logged to the program's output.

The main thing (to me) is to get the error reporting working, but we do want to know what the underlying cause of your problem is.

**Comment by Gamerpunk25** [comment]
Makes sense for me. Without a correct working error reporting, how should anybody know, what the flaw is. That's like searching for the illness but you can't see the symptoms. I will give it a try now and report, what is happening

**Comment by danieljohnson2** [comment]
Rats.

Your Lutris is installed as a Flatpak, which provides all the dependencies with in. But what you downloaded is just the Lutris source tree, which does nothing of the kind. I was hoping you had enough stuff installed to try this. We can't get at the Flatpak stuff for this test, that's why it's failing, I think.

Can you install any version of Lutris with your package manager, temporarily? That should get all the dependencies we need installed.

**Comment by Gamerpunk25** [comment]
After the completed installation of the lutris tuxedo version and then starting the provided script from your branch, the gui showed up with the sources GOG, Epic, EA App, Ubisoft and Steam on the side. After i added amazon prime gaming as a source via preferences and logged in, the following error message showed up:

HTTP Error 504: Gateway Time-out

Traceback (most recent call last):

  File "/home/mordecai/Downloads/lutris-dj-amazon_error_handling/lutris/util/http.py", line 97, in _request
    request = urllib.request.urlopen(req, timeout=self.timeout)  # pylint: disable=consider-using-with

  File "/usr/lib/python3.10/urllib/request.py", line 216, in urlopen
    return opener.open(url, data, timeout)

  File "/usr/lib/python3.10/urllib/request.py", line 525, in open
    response = meth(req, response)

  File "/usr/lib/python3.10/urllib/request.py", line 634, in http_response
    response = self.parent.error(

  File "/usr/lib/python3.10/urllib/request.py", line 563, in error
    return self._call_chain(*args)

  File "/usr/lib/python3.10/urllib/request.py", line 496, in _call_chain
    result = func(*args)

  File "/usr/lib/python3.10/urllib/request.py", line 643, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)

urllib.error.HTTPError: HTTP Error 504: Gateway Time-out


The above exception was the direct cause of the following exception:


Traceback (most recent call last):

  File "/home/mordecai/Downloads/lutris-dj-amazon_error_handling/lutris/util/jobs.py", line 30, in target
    result = self.function(*a, **kw)

  File "/home/mordecai/Downloads/lutris-dj-amazon_error_handling/lutris/services/base.py", line 143, in do_reload
    self.load()

  File "/home/mordecai/Downloads/lutris-dj-amazon_error_handling/lutris/services/amazon.py", line 163, in load
    games = [AmazonGame.new_from_amazon_game(game) for game in self.get_library()]

  File "/home/mordecai/Downloads/lutris-dj-amazon_error_handling/lutris/services/amazon.py", line 356, in get_library
    json_data = self.request_sds(

  File "/home/mordecai/Downloads/lutris-dj-amazon_error_handling/lutris/services/amazon.py", line 431, in request_sds
    request.post(json.dumps(body).encode())

  File "/home/mordecai/Downloads/lutris-dj-amazon_error_handling/lutris/util/http.py", line 140, in post
    return self._request("POST", data)

  File "/home/mordecai/Downloads/lutris-dj-amazon_error_handling/lutris/util/http.py", line 101, in _request
    raise HTTPError("%s" % error, code=error.code) from error

lutris.util.http.HTTPError: HTTP Error 504: Gateway Time-out

Lutris log:
[INFO:2024-09-10 17:09:53,719:application]: Starting Lutris 0.5.17
[INFO:2024-09-10 17:09:53,822:__init__]: Running migration: mess_to_mame
[INFO:2024-09-10 17:09:53,826:__init__]: Running migration: migrate_hidden_ids
[INFO:2024-09-10 17:09:53,828:__init__]: Running migration: migrate_steam_appids
[INFO:2024-09-10 17:09:53,831:__init__]: Running migration: migrate_banners
[INFO:2024-09-10 17:09:53,833:__init__]: Running migration: retrieve_discord_appids
[INFO:2024-09-10 17:09:53,835:retrieve_discord_appids]: Updating Games Discord APP ID's
[INFO:2024-09-10 17:09:53,837:__init__]: Running migration: migrate_sortname
[INFO:2024-09-10 17:09:53,838:migrate_sortname]: Adding blank sortname field to database
[INFO:2024-09-10 17:09:53,840:__init__]: Running migration: migrate_hidden_category
[INFO:2024-09-10 17:09:53,842:migrate_hidden_category]: Moving hidden games to the '.hidden' category
[INFO:2024-09-10 17:09:56,553:startup]: NVIDIA GeForce RTX 4060 Laptop GPU (10de:28a0 1d05:128a nvidia) Driver 550.90.07
[INFO:2024-09-10 17:09:57,439:startup]: AMD Radeon Graphics (1002:15bf 1d05:134f amdgpu) Driver 24.0.8
[WARNING:2024-09-10 17:10:17,480:system]: Non existent path: /home/mordecai/.local/share/lutris/runtime/battleye_runtime
[WARNING:2024-09-10 17:10:17,764:system]: Non existent path: /home/mordecai/.local/share/lutris/runtime/umu
[WARNING:2024-09-10 17:10:19,001:system]: Non existent path: /home/mordecai/.local/share/lutris/runtime/umu-games
[WARNING:2024-09-10 17:10:24,366:system]: Non existent path: /home/mordecai/.local/share/lutris/runtime/eac_runtime
[WARNING:2024-09-10 17:10:24,683:system]: Non existent path: /home/mordecai/.local/share/lutris/runtime/winetricks
[WARNING:2024-09-10 17:10:41,728:system]: Non existent path: /home/mordecai/.local/share/lutris/runtime/innoextract-1.9-linux
[WARNING:2024-09-10 17:10:44,174:system]: Non existent path: /home/mordecai/.local/share/lutris/runtime/p7zip
[WARNING:2024-09-10 17:10:45,153:system]: Non existent path: /home/mordecai/.local/share/lutris/runtime/gamecontrollerdb
[WARNING:2024-09-10 17:10:46,714:system]: Non existent path: /home/mordecai/.local/share/lutris/runtime/network-tools
[WARNING:2024-09-10 17:11:01,878:system]: Non existent path: /home/mordecai/.local/share/lutris/runtime/Ubuntu-18.04-i686
[WARNING:2024-09-10 17:11:17,004:system]: Non existent path: /home/mordecai/.local/share/lutris/runtime/Ubuntu-18.04-x86_64
[WARNING:2024-09-10 17:11:18,259:system]: Non existent path: /home/mordecai/.local/share/lutris/runtime/dxvk-cache-tool
[INFO:2024-09-10 17:14:57,578:amazon]: Generated code_verifier: b'chKqIIriGv5hFNfcuVThNO1RQQDLUc0oaohGqgum4wg'
[INFO:2024-09-10 17:14:57,579:amazon]: Generated challange: b'fLD8p5b92KznmtpLmAXf8pQKmmWfD_e4PbQ9jUZrS4Y'
[INFO:2024-09-10 17:14:57,837:amazon]: Generated serial: 00000000000000000000B99973C64483
[INFO:2024-09-10 17:14:57,837:amazon]: Generated client_id: 3030303030303030303030303030303030303030423939393733433634343833234132554d56484f58375550345637
[INFO:2024-09-10 17:15:50,247:amazon]: Got authorization code
[INFO:2024-09-10 17:15:50,249:amazon]: Registerring a device. ID: 3030303030303030303030303030303030303030423939393733433634343833234132554d56484f58375550345637
[INFO:2024-09-10 17:15:51,251:amazon]: Successfully registered a device
[ERROR:2024-09-10 17:15:58,169:jobs]: Error while completing task <function BaseService.start_reload.<locals>.do_reload at 0x7748700d70a0>: <class 'lutris.util.http.HTTPError'> HTTP Error 504: Gateway Time-out

**Comment by danieljohnson2** [comment]
OK, that's progress. I can merge the error reporting at least.

But we can now see that you are getting 504 Gateway Timeout when you connect to Amazon. It is trying to reach ``https://sds.amazon.com/amazon/`` I think. I can you reach that URL with a web browser, at all? It's not a web page, you'll just get an error there, but is it a 504 Gateway Timeout?

**Comment by Gamerpunk25** [comment]
I typed this address in firefox and tried it. In addition, i let the network analysis of firefox do its job during the search for this address. According to the result, i get the html code 500, standing for internal server error. Below the screenshot
![grafik](https://github.com/user-attachments/assets/246f96da-2e7e-4dff-a9f7-9bc22ed15b99)

**Comment by danieljohnson2** [comment]
That's what I'm getting too. Somehow Lutris gets 504 though.

I may need alter Lutris again to get the exact URL we're using here; but I can't do it now, so I'll get back to you, again.

**Comment by Gamerpunk25** [comment]
At least the error reporting goes now. I'll wait for the next attempt

**Comment by strycore** [comment]
https://discord.com/channels/512538904872747018/799769240471207996/1281360517096673445

**Comment by Gamerpunk25** [comment]
Why this discord link all of a sudden?

**Comment by strycore** [comment]
has the fix

**Comment by Gamerpunk25** [comment]
This link does open discord, but on the sidebar, i can only see the servers i am invited. Nothing else like channels or a chat

**Comment by strycore** [comment]
https://github.com/imLinguin/nile/commit/aefa5dd9c3a5146669da22317c8e0e3c12665f64

**Comment by danieljohnson2** [comment]
I think @strycore is saying that 0.5.17 can't connect to Amazon because they changed their protocols. PR #5528 was supposed to be the fix, and it's merged, so you should be able to reload using the branch I gave you. But apparently not.

Maybe that PR did not do the trick, or maybe something else is wrong.

**Comment by danieljohnson2** [comment]
I've merged the branch; at least it fixes the bogus NoneType error.

But I'm pretty sure the URL being queried here is just ``https://sds.amazon.com/amazon/``, so what I don't understand is why Lutris gets 504 here when Firefox does not; maybe it depends on whether its a POST or a GET?

I don't know if the Amazon source is working for anyone- I don't have an account to check with- but my gut instinct is that this is some kind of networking issue preventing you from reaching their server in the first place.

**Comment by Gamerpunk25** [comment]
So that i get the current state straight, for me using amazon gaming is the Nile project the better solution, until the sync with Amazon works properly. And what further investigation could help for lutris?

**Comment by danieljohnson2** [comment]
Could you try Nile out? I'd like to know if it works for you.

**Comment by Gamerpunk25** [comment]
i tried the Nile project at first. But it doesn't want to register my device. it's is a simple program to run trough the command line. i'll have to read further into it. Maybe it turns out as a simple replacement but i am thrilled to see, that this problem in lutris can be solved

**Comment by danieljohnson2** [comment]
If Lutris fails, but Nile works,  that suggests there's something wrong with the Lutris implementation here. But it's not really the same bug as the NoneType thing. I would say it should be an new issue.

However, I hope that isn't true, since I can't really fix this. I don't have a suitable Amazon account, and this is more than I can do without being about to test it myself.

**Comment by danieljohnson2** [comment]
@Gamerpunk25, can you try PR #5598? It's supposed to fix the 504 failure...

**Comment by Gamerpunk25** [comment]
I will try PR #5598 and report what happened

**Comment by Gamerpunk25** [comment]
I replaced the referred file of #5598 in the branch, which prevents bogus error handling. After reloading the amazon game library, i get an overview of my games. No error message or such.

**Comment by danieljohnson2** [comment]
OK, I guess this is a fix. I'll merge it.

**Comment by Gamerpunk25** [comment]
perfect. With this merge, this issue can be considered as closed

**Comment by Felzow47** [comment]
still have same error installed from discover
![image](https://github.com/user-attachments/assets/3d57ba61-a674-454a-b4eb-25eef99812cd)

**Comment by Gamerpunk25** [comment]
the current release 0.5.17 doesn't has this fix. the merge happened in the master branch. the cloned master branch doesn't contain this issue. You could clone this repository via git and start this

**Comment by WellBeSerious** [comment]
This is fixed in #5612 and #5613 by downloading and running Master. Make sure Lutris dependencies are installed (like by having Lutris installed before).
